promql/binary_expr.go

```diff
@@ -0,0 +1,387 @@
+	return &ast.FunctionExpression{
+		Params: []*ast.Property{
+			{
+				Key: &ast.Identifier{
+					Name: "column",
+				},
+			},
+		},
+		Body: call("strings.trimSuffix", map[string]ast.Expression{
+			"v":      &ast.Identifier{Name: "column"},
+			"suffix": &ast.StringLiteral{Value: suffix},
+		}),
+	}
+}
+
+// Function to match post-join RHS columns.
+var matchRHSSuffixFn = &ast.FunctionExpression{
+	Params: []*ast.Property{
+		{
+			Key: &ast.Identifier{
+				Name: "column",
+			},
+		},
+	},
+	Body: &ast.BinaryExpression{
+		Operator: ast.RegexpMatchOperator,
+		Left:     &ast.Identifier{Name: "column"},
+		Right:    &ast.RegexpLiteral{Value: regexp.MustCompile("_rhs$")},
+	},
+}
+
+func (t *Transpiler) transpileBinaryExpr(b *promql.BinaryExpr) (ast.Expression, error) {
+	lhs, err := t.transpileExpr(b.LHS)
+	if err != nil {
+		return nil, fmt.Errorf("unable to transpile left-hand side of binary operation: %s", err)
+	}
+	rhs, err := t.transpileExpr(b.RHS)
+	if err != nil {
+		return nil, fmt.Errorf("unable to transpile right-hand side of binary operation: %s", err)
+	}
+
+	swapped := false
+
+	switch {
+	case yieldsFloat(b.LHS) && yieldsFloat(b.RHS):
+		if op, ok := arithBinOps[b.Op]; ok {
+			return &ast.BinaryExpression{
+				Operator: op,
+				Left:     lhs,
+				Right:    rhs,
+			}, nil
+		}
+
+		if opFn, ok := arithBinOpFns[b.Op]; ok {
+			return call(opFn, map[string]ast.Expression{"x": lhs, "y": rhs}), nil
+		}
+
+		if op, ok := compBinOps[b.Op]; ok {
+			if !b.ReturnBool {
+				// This is already caught by the PromQL parser.
+				panic("scalar-to-scalar binary op is missing 'bool' modifier")
+			}
+
+			return call("float", map[string]ast.Expression{
+				"v": &ast.BinaryExpression{
+					Operator: op,
+					Left:     lhs,
+					Right:    rhs,
+				},
+			}), nil
+		}
+
+		panic(fmt.Errorf("invalid scalar-scalar binary op %q", b.Op))
+	case yieldsFloat(b.LHS) && yieldsTable(b.RHS):
+		lhs, rhs = rhs, lhs
+		swapped = true
+		fallthrough
+	case yieldsTable(b.LHS) && yieldsFloat(b.RHS):
+		if op, ok := arithBinOps[b.Op]; ok {
+			return buildPipeline(
+				lhs,
+				call("map", map[string]ast.Expression{"fn": scalarArithBinaryOpFn(op, rhs, swapped)}),
+				dropFieldAndTimeCall,
+			), nil
+		}
+
+		if opFn, ok := arithBinOpFns[b.Op]; ok {
+			return buildPipeline(
+				lhs,
+				call("map", map[string]ast.Expression{"fn": scalarArithBinaryMathFn(opFn, rhs, swapped)}),
+				dropFieldAndTimeCall,
+			), nil
+		}
+
+		if op, ok := compBinOps[b.Op]; ok {
+			if b.ReturnBool {
+				return buildPipeline(
+					lhs,
+					call("map", map[string]ast.Expression{
+						"fn": scalarArithBinaryOpFn(op, rhs, swapped),
+					}),
+					call("toFloat", nil),
+					dropFieldAndTimeCall,
+				), nil
+			}
+			return buildPipeline(
+				lhs,
+				call("filter", map[string]ast.Expression{"fn": scalarCompBinaryOpFn(op, rhs, swapped)}),
+			), nil
+		}
+
+		panic(fmt.Errorf("invalid scalar-vector binary op %q", b.Op))
```

Comment:
Yes, there's a couple of cases where I rely on the PromQL parser to already verify things and panic if that assumption is violated (like here). I'll convert those panics to errors then.

Note that there'll still be some places where I make those assumptions implicitly without an explicit check&panic for everything.

Reply:
(Hmm, keeping the panics in the label matchers transpilation for now as it would otherwise blow up the code needlessly, since there's no errors expected in that code path so far. But converted the other instances).